### PR TITLE
[1/3][Distributed] Make distributed thunks the witnesses, fix calls on generic DAs

### PR DESCRIPTION
**Pull request by ktoso**
Implements calls on generic distributed actors, by fixing how the witness is found: it must be the thunk.

Resolves rdar://95829405

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please build toolchain

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4791,6 +4791,9 @@ ERROR(distributed_actor_func_static,none,
 ERROR(distributed_actor_func_not_in_distributed_actor,none,
       "'distributed' method can only be declared within 'distributed actor'",
       ())
+ERROR(distributed_method_requirement_must_be_async_throws,none, // FIXME(distributed): this is an implementation limitation we should lift
```
Comment (ktoso) [inline include/swift/AST/DiagnosticsSema.def:4794]: marked all sites where we emit this with rdar://95949498

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/CodeSynthesisDistributedActor.cpp [hunk 2]
```diff
@@ -648,8 +648,13 @@ static FuncDecl *createDistributedThunkFunction(FuncDecl *func) {
   auto &C = func->getASTContext();
   auto DC = func->getDeclContext();
 
-  auto systemTy = getConcreteReplacementForProtocolActorSystemType(func);
-  assert(systemTy &&
+  // NOTE: So we don't need a thunk in the protocol, we should call the underlying
+  // thing instead, which MUST have a thunk, since it must be a distributed func as well...
+  if (dyn_cast<ProtocolDecl>(DC)) {
```
Comment (DougGregor) [inline lib/Sema/CodeSynthesisDistributedActor.cpp:653]: Nit: `s/dyn_cast/isa/` (not important)

File: lib/SILGen/SILGenPoly.cpp [hunk 3]
```diff
@@ -4439,15 +4439,22 @@ void SILGenFunction::emitProtocolWitness(
...
+      getActorIsolation(witness.getDecl()).isDistributedActor()) {
+    // We witness protocol requirements using the distributed thunk, when:
+    // - the witness is isolated to a distributed actor, but the requirement is not
+    // - the requirement is a distributed func, and therefore can only be witnessed
+    //   by a distributed func; we handle this by witnessing the requirement with the thunk
+    // FIXME(distributed): this limits us to only allow distributed explicitly throwing async requirements... we need to fix this somehow.
+    if (requirement.hasDecl()) {
+      if ((!getActorIsolation(requirement.getDecl()).isDistributedActor()) ||
+          (isa<FuncDecl>(requirement.getDecl()) &&
+              witness.getFuncDecl()->isDistributed())) {
```
Comment (DougGregor) [inline lib/SILGen/SILGenPoly.cpp:4452]: I think the last line was meant to be `requirement.getFuncDecl()->isDistributed()`, i.e., is the requirement distributed.

**Review by DougGregor — APPROVED** [review]
Minor nits here, but overall this is an improvement and the type checker behavior accurately reflects the implementation behavior.

**Comment by ktoso** [comment]
Thanks for review, will address in the 3/3 PR 👍

File: lib/SILGen/SILGenPoly.cpp (hunk 3 above)
Comment (ktoso) [reply lib/SILGen/SILGenPoly.cpp:4452]: Ah yes, thank you!

Comment (ktoso) [reply lib/SILGen/SILGenPoly.cpp:4452]: Fixed in https://github.com/apple/swift/pull/59722

**Pull request merged**